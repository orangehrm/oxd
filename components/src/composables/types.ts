import {InjectionKey, Ref} from 'vue';

export type ErrorField = {
  cid: string;
  errors: string[];
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Rule = (value?: any) => string | boolean;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ModelValue = Ref<any>;

export type ErrorBag = Array<ErrorField>;
export type Fieldset = Array<FormField>;

export interface FormField {
  cid: Ref<string>;
  label: Ref<string>;
  dirty: Ref<boolean>;
  touched: Ref<boolean>;
  processing: Ref<boolean>;
  validate: () => Promise<ErrorField>;
  reset: () => void;
}

export interface FormAPI {
  searchErrors: (id: string) => ErrorField[];
  purgeErrors: (id: string) => void;
  addError: (errorField: ErrorField) => void;
  registerField: (field: FormField) => void;
  unregisterField: (field: FormField) => void;
}

export const formKey: InjectionKey<FormAPI> = Symbol('form');
