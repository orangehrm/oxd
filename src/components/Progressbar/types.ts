/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

export const TYPE_DEFAULT = 'default';
export const TYPE_MAIN = 'main';
export const TYPE_SECONDARY = 'secondary';
export const TYPE_SUCCESS = 'success';
export const TYPE_WARN = 'warn';
export const TYPE_ERROR = 'error';
export const TYPE_INFO = 'info';

export const TYPES = [
  TYPE_DEFAULT,
  TYPE_MAIN,
  TYPE_SECONDARY,
  TYPE_SUCCESS,
  TYPE_WARN,
  TYPE_ERROR,
  TYPE_INFO,
];

export type ProgressbarType =
  | typeof TYPE_DEFAULT
  | typeof TYPE_MAIN
  | typeof TYPE_SECONDARY
  | typeof TYPE_SUCCESS
  | typeof TYPE_WARN
  | typeof TYPE_ERROR
  | typeof TYPE_INFO;
