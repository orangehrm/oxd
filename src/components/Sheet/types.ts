/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

export const TYPE_GRAY_DARKEN_2 = 'gray-darken-2';
export const TYPE_GRAY_DARKEN_1 = 'gray-darken-1';
export const TYPE_GRAY = 'gray';
export const TYPE_GRAY_LIGHTEN_1 = 'gray-lighten-1';
export const TYPE_GRAY_LIGHTEN_2 = 'gray-lighten-2';
export const TYPE_LIGHT_GRAY = 'light-gray';
export const TYPE_PASTEL_WHITE = 'pastel-white';
export const TYPE_WHITE_SHADOW = 'white-shadow';
export const TYPE_WHITE = 'white';

export const TYPES = [
  TYPE_GRAY_DARKEN_2,
  TYPE_GRAY_DARKEN_1,
  TYPE_GRAY,
  TYPE_GRAY_LIGHTEN_1,
  TYPE_GRAY_LIGHTEN_2,
  TYPE_LIGHT_GRAY,
  TYPE_PASTEL_WHITE,
  TYPE_WHITE_SHADOW,
  TYPE_WHITE,
];

export type SheetType =
  | typeof TYPE_GRAY_DARKEN_2
  | typeof TYPE_GRAY_DARKEN_1
  | typeof TYPE_GRAY
  | typeof TYPE_GRAY_LIGHTEN_1
  | typeof TYPE_GRAY_LIGHTEN_2
  | typeof TYPE_LIGHT_GRAY
  | typeof TYPE_PASTEL_WHITE
  | typeof TYPE_WHITE_SHADOW
  | typeof TYPE_WHITE;
