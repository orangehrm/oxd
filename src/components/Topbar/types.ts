/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

export interface TopMenuItem {
  name: string;
  url: string;
  active: boolean;
  children: SubMenuItem[];
}

export interface SubMenuItem {
  name: string;
  url: string;
}

export interface User {
  firstName: string;
  lastName: string;
  profImgSrc: string;
}

export interface Breadcrumb {
  moduleName: string;
  level?: string;
}
