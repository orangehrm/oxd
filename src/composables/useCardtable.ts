/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {injectStrict} from '../utils/injectable';
import type {ResponsiveState} from './useResponsive';
import {provide, readonly, type InjectionKey} from 'vue';

export type TableProps = {
  [key: string]: any;
};

export const tablePropsKey: InjectionKey<TableProps> = Symbol('tableProps');
export const tableScreenStateKey: InjectionKey<ResponsiveState> =
  Symbol('screenState');

export const useProvideTableProps = (
  props: TableProps,
  screenState: ResponsiveState,
) => {
  provide(tablePropsKey, readonly(props));
  provide(tableScreenStateKey, readonly(screenState));
};

export const useInjectTableProps = () => {
  const tableProps = injectStrict<TableProps>(tablePropsKey);
  const screenState = injectStrict<ResponsiveState>(tableScreenStateKey);

  return {
    tableProps,
    screenState,
  };
};
