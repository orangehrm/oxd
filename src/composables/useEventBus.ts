/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

interface EventQueue {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: Array<(data: any) => void>;
}

class EventBus {
  private queue: EventQueue;
  constructor() {
    this.queue = {};
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  $on(name: string, callback: (data: any) => void) {
    this.queue[name] = this.queue[name] || [];
    this.queue[name].push(callback);
  }

  $off(name: string, callback: () => void) {
    if (this.queue[name]) {
      for (let i = 0; i < this.queue[name].length; i++) {
        if (this.queue[name][i] === callback) {
          this.queue[name].splice(i, 1);
          break;
        }
      }
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  $emit(name: string, data: any) {
    if (this.queue[name]) {
      this.queue[name].forEach(function (callback) {
        callback(data);
      });
    }
  }
}

export default function useEventBus() {
  return new EventBus();
}
