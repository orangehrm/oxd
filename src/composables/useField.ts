/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {nanoid} from 'nanoid';
import {formKey} from './types';
import type {Ref, WatchStopHandle} from 'vue';
import {injectStrict} from '../utils/injectable';
import type {ErrorField, FormAPI, Rule} from './types';
import {ref, watch, computed, onBeforeUnmount} from 'vue';

export default function useField(fieldContext: {
  fieldLabel: string;
  rules: Rule[];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  modelValue: Ref<any>;
  onReset: () => Promise<void>;
  disabled: Ref<boolean>;
}) {
  const form = injectStrict<FormAPI>(formKey);
  const cid = ref<string>(nanoid());
  const label = ref<string>(fieldContext.fieldLabel);
  const dirty = ref<boolean>(false);
  const touched = ref<boolean>(false);
  const processing = ref<boolean>(false);
  let watchHandler: WatchStopHandle;

  const validate = () => {
    if (fieldContext.disabled.value) {
      return Promise.resolve({cid: cid.value, errors: []});
    }

    async function allValidations() {
      for (const rule of fieldContext.rules) {
        const validationOutput = await rule(fieldContext.modelValue.value);
        if (validationOutput === true) continue;
        return typeof validationOutput === 'string'
          ? validationOutput
          : `Rules should return a string or true, received '${typeof validationOutput}'`;
      }
    }

    return new Promise<ErrorField>((resolve) => {
      processing.value = true;
      allValidations()
        .then((resolved) => {
          resolve({
            cid: cid.value,
            errors: typeof resolved === 'string' ? [resolved] : [],
          });
        })
        .finally(() => {
          processing.value = false;
        });
    });
  };

  const startWatcher = () => {
    watchHandler = watch(fieldContext.modelValue, () => {
      validate().then((result) => {
        form.addError(result);
      });
    });
  };

  startWatcher();

  const reset = () => {
    dirty.value = false;
    touched.value = false;
    processing.value = false;
    watchHandler(); // stop the validation watcher
    fieldContext.onReset().then(() => startWatcher());
  };

  form.registerField({cid, label, dirty, touched, processing, validate, reset});

  onBeforeUnmount(() => {
    form.unregisterField({
      cid,
      label,
      dirty,
      touched,
      processing,
      validate,
      reset,
    });
  });

  const hasError = computed(() => {
    return form.searchErrors(cid.value).length > 0;
  });

  const message = computed(() => {
    return hasError.value ? form.searchErrors(cid.value)[0].errors[0] : null;
  });

  return {
    form,
    reset,
    validate,
    hasError,
    message,
  };
}
