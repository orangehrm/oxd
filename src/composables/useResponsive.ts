/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {nextTick, onBeforeUnmount, onMounted, reactive} from 'vue';

export const DEVICE_XS = 'xs';
export const DEVICE_SM = 'sm';
export const DEVICE_MD = 'md';
export const DEVICE_LG = 'lg';
export const DEVICE_XL = 'xl';

interface Breakpoint {
  [screenType: string]: {
    min: number;
    max: number;
  };
}

export interface ResponsiveState {
  windowWidth: number;
  windowHeight: number;
  screenType: string;
}

export const breakpoints: Breakpoint = {
  [DEVICE_XS]: {
    min: 0,
    max: 599,
  },
  [DEVICE_SM]: {
    min: 600,
    max: 799,
  },
  [DEVICE_MD]: {
    min: 800,
    max: 999,
  },
  [DEVICE_LG]: {
    min: 1000,
    max: 1199,
  },
  [DEVICE_XL]: {
    min: 1200,
    max: 4000,
  },
};

export default function useResponsive() {
  const state: ResponsiveState = reactive({
    windowWidth: 0,
    windowHeight: 0,
    screenType: DEVICE_XS,
  });

  const setWindowSize = () => {
    const width = window.innerWidth;
    const height = window.innerHeight;
    state.windowWidth = width;
    state.windowHeight = height;
    for (const screenType in breakpoints) {
      if (
        breakpoints[screenType].min <= width &&
        breakpoints[screenType].max >= width
      ) {
        state.screenType = screenType;
        break;
      }
    }
  };

  onMounted(async () => {
    await nextTick();
    setWindowSize();
    window.addEventListener('resize', setWindowSize);
  });

  onBeforeUnmount(() => {
    window.removeEventListener('resize', setWindowSize);
  });

  return state;
}
