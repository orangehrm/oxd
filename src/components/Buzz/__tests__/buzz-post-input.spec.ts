/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {mount} from '@vue/test-utils';
import {formKey} from '@/composables/types';
import type {FormAPI} from '@/composables/types';
import {describe, expect, it, vi} from 'vitest';
import BuzzPostInput from '@/components/Buzz/BuzzPostInput.vue';

describe('BuzzPostInput.vue', () => {
  const mockFormAPI: FormAPI = {
    searchErrors: vi.fn(() => []),
    purgeErrors: vi.fn(),
    addError: vi.fn(),
    registerField: vi.fn(),
    unregisterField: vi.fn(),
  };
  it('renders OXD BuzzPostInput', () => {
    const wrapper = mount(BuzzPostInput, {
      props: {},
      global: {
        provide: {
          [formKey as symbol]: mockFormAPI,
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
