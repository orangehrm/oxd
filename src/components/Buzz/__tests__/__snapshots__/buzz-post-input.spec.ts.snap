// Vitest Snapshot v1

exports[`BuzzPostInput.vue > renders OXD BuzzPostInput 1`] = `
"<div class=\\"oxd-buzz-post oxd-buzz-post--active\\" data-v-ec69d519=\\"\\"><textarea rows=\\"1\\" class=\\"oxd-buzz-post-input\\" data-v-ec69d519=\\"\\"></textarea>
  <div class=\\"oxd-buzz-post-slot\\" data-v-ec69d519=\\"\\"></div>
</div>
<!--v-if-->"
`;
